File: es5/problems/p023.js
var number = require('../utils/number');

var abundant = [];
var isAbundant = {};
var result = 0;
var i;

var sumDivisors = function(num) {
    var divisors = number.properDivisors(num);
    var result = 0;

    divisors.pop();

    for (var i = 0; i < divisors.length; i++) {
        result += divisors[i];
    }
    
    return result;
};

var isSumOfTwoAbundants = function(num) {
    for (var i = 0; i < abundant.length; i++) {
        if (abundant[i] >= num) {
            return false;
        }
        if (isAbundant[num - abundant[i]] === true) {
            return true;
        }
    }
    return false;
};

for (i = 1; i < 28123; i++) {
    if (sumDivisors(i) > i) {
        isAbundant[i] = true;
        abundant.push(i);
    }
}

abundant.sort(function(a,b) { return a - b; });

for (i = 1; i < 28123; i++) {
    if (!isSumOfTwoAbundants(i)) {
        result += i;
    }
}

console.log(result);

File: rust/p023.rs
use std::collections::HashSet;
use std::iter::FromIterator;
mod primes;
mod numbers;
use numbers::Factor;

fn sum_divisors(primes: &[usize], number: usize) -> usize {
    let factors = Factor::factorize(primes, number);
    let mut divisors = Factor::proper_divisors(&factors);
    divisors.pop();
    divisors.iter().sum()
}

fn is_sum_of_two_abundants(abundant: &[usize], abundant_set: &HashSet<&usize>, number: usize) -> bool {
    for &current in abundant {
        if current >= number {
            return false;
        }

        let pair = number - current;
        if abundant_set.contains(&pair) {
            return true;
        }
    }
    false
}

fn main() {
	let primes = primes::eratosthenes(100000);

    let abundant: Vec<usize> = (1..28123).into_iter()
        .filter(|&x| sum_divisors(&primes, x) > x)
        .collect::<Vec<usize>>();

    let abundant_set = HashSet::from_iter(abundant.iter());

    let result: usize = (1..28123).into_iter()
        .filter(|&x| !is_sum_of_two_abundants(&abundant, &abundant_set, x))
        .sum();
    println!("{}", result);
}
